### ISSUE TITLE
If an error is not affected by a macro, don't print macro notes

### ISSUE THREAD

**Issue by nrc** [rust-lang/rust#30506]
e.g.,

```
foo!(fasdfas);
```

If there is an error about `fasdfas` it should just use the span for `fasdfas`, and not print anything about `foo!`.

cc @DanielJCampbell

**Comment by durka** [comment]
Hmm what do you mean by "an error about `fasdfas`"? Definitely the macro backtrace adds a lot of noise to the error message, but isn't the macro free to reinterpret `fasdfas` in various ways so how is the compiler supposed to know whether `foo!` is relevant or not?

**Comment by nrc** [comment]
The compiler sees things somewhat backwards - it sees the post-expansion AST and an expansion trace. Generated code, can come from three places - straight from the macro, straight from the input, or some complex mashing up of the two via syntax extensions or some of the built-in macros. For the first and third cases if there is an error, we can't do much better than giving the whole instantiation (e.g., `foo!(fasdfas)`) and hopefully some part of the macro definition (in some cases, perhaps only the whole macro definition, I'm not sure). However, in the middle case, the expansion trace shows that the macro is just copying and pasting it's input into somewhere in the macro. The compiler should be able to tell easily if that is the case from the expansion trace (and @DanielJCampbell has been doing some work improving these traces in any case), and then skip informing the user about the macro stuff, since it doesn't matter too much (we might still note that a macro was involved and reference it, but show less detail than we do at the moment).

**Comment by nrc** [comment]
Actually, we seem to be doing much better than we used - we don't have the annoying `note: expansion site` notes any more.

I still think we should consider dropping any mention of the macro - after all if there is a problem with a sub-expression in a function call, we don't mention the function.

Examples:
```
macro_rules! foo{
    ($x: expr) => {$x}
}

fn main() {
    foo!(bar());
}
```

would give just

```
<anon>:6:10: 6:13 error: unresolved name `bar` [E0425]
<anon>:6     foo!(bar());
                  ^~~
<anon>:6:10: 6:13 help: see the detailed explanation for E0425
```
losing (because it is not useful)

```
<anon>:6:5: 6:17 note: in this expansion of foo! (defined in <anon>)
```

but
```
macro_rules! foo{
    ($x: expr) => {$x.bar()}
}

fn main() {
    foo!(42);
}
```
would stay as it is:

```
<anon>:2:23: 2:28 error: no method named `bar` found for type `_` in the current scope
<anon>:2     ($x: expr) => {$x.bar()}
                               ^~~~~
<anon>:6:5: 6:14 note: in this expansion of foo! (defined in <anon>)
```

**Comment by jseyfried** [comment]
Fixed in #40597.
